src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs

```diff
@@ -357,8 +375,10 @@ public void RemoveListener(Action<IMessageSource, TMessage> listenerFunc)
             ListenerDict = new ConcurrentDictionary<Type, Dispatcher>();
             ProgressTracker = new ProgressReporting.ProgressTracker(this);
             _cancelLock = new object();
+            _cancelEnvLock = new object();
```

Comment:
This "smells" to me that we have both `_cancelLock` and `_cancelEnvLock`.

I think it would make more sense if we moved the `HostBase._children` field "up" to the HostEnvironmentBase class. Then we only need a single `_cancelLock` object. And we only need to a single "children" list.

Reply:
You have a good smell.

---
In reply to: [266936547](https://github.com/dotnet/machinelearning/pull/2797#discussion_r266936547) [](ancestors = 266936547)